### ISSUE TITLE
floating invalid of F2010C5-CMIP6-LR.ne30_oECv3 on cori-knl with maint-1.0

### ISSUE THREAD

**Issue by tangq** [E3SM-Project/E3SM#3328]
I got the following floating invalid errors when trying to rerun one of my low-res cases on cori-knl with current maint-1.0.

I reproduced this error with multiple submissions. Any suggestions? Thanks.

create_newcase options are `--compset F2010C5-CMIP6-LR --res ne30_oECv3 --pecount L --mach cori-knl`
$rundir: /global/cscratch1/sd/tang30/E3SM_simulations/cori-knl.20191124.tst2.F2010C5-CMIP6-LR.ne30_oECv3

3350: forrtl: error (65): floating invalid
3350: Image              PC                Routine            Line        Source             
3350: e3sm.exe           0000000004329BCE  Unknown               Unknown  Unknown
3350: e3sm.exe           0000000003B96D00  Unknown               Unknown  Unknown
3350: e3sm.exe           000000000107EAE3  clubb_intr_mp_clu        1584  clubb_intr.F90
3350: e3sm.exe           00000000005FA3DB  physpkg_mp_tphysb        2479  physpkg.F90
3350: e3sm.exe           00000000005F6B12  physpkg_mp_phys_r        1036  physpkg.F90
3350: e3sm.exe           00000000004E79E7  cam_comp_mp_cam_r         250  cam_comp.F90
3350: e3sm.exe           00000000004D9CAC  atm_comp_mct_mp_a         341  atm_comp_mct.F90
3350: e3sm.exe           000000000042367F  component_mod_mp_         269  component_mod.F90
3350: e3sm.exe           000000000041768D  cime_comp_mod_mp_        1958  cime_comp_mod.F90
3350: e3sm.exe           000000000042042C  MAIN__                     92  cime_driver.F90
3350: e3sm.exe           0000000000401ADE  Unknown               Unknown  Unknown
3350: e3sm.exe           000000000440968F  Unknown               Unknown  Unknown
3350: e3sm.exe           00000000004019CA  Unknown               Unknown  Unknown

**Comment by singhbalwinder** [comment]
@tangq: Would you please try @ndkeen 's solution described in #3324 to see if it helps?

**Comment by tangq** [comment]
Thanks for the suggestion, @singhbalwinder .

I am testing now..will update later.

**Comment by ndkeen** [comment]
Well the change in PR 3324 will likely only affect DEBUG=TRUE runs. Qi's error might be something else. Can you explain any more? Is this maybe an older repo you are trying to use?

**Comment by singhbalwinder** [comment]
@ndkeen : I was also surprised at that but it seems like @tangq seems to be using a few debug flags which can capture this error. I am wondering what impact that might have on the performance of the code. It should definitely slow down the simulation.

If these flags are being used in the default production runs, we should change that in the machine files so that we improve the throughput.

**Comment by tangq** [comment]
@ndkeen , @singhbalwinder , I tried the fix in #3324 and the test failed with floating invalid error too, but seems at a different place. Errors are pasted below.

I am not using DEBUG=TRUE, instead my settings are default for production runs. I used the latest maint-1.0 code. Please let me know if you need further information to solve this issue.

`2546: forrtl: error (65): floating invalid
2546: Image              PC                Routine            Line        Source             
2546: e3sm.exe           0000000004329BCE  Unknown               Unknown  Unknown
2546: e3sm.exe           0000000003B96D00  Unknown               Unknown  Unknown
2546: e3sm.exe           0000000000F7592C  aer_rad_props_mp_         245  aer_rad_props.F90
2546: e3sm.exe           000000000064D47A  radiation_mp_radi        1214  radiation.F90
2546: e3sm.exe           00000000005FE2B0  physpkg_mp_tphysb        2692  physpkg.F90
2546: e3sm.exe           00000000005F6B12  physpkg_mp_phys_r        1036  physpkg.F90
2546: e3sm.exe           00000000004E79E7  cam_comp_mp_cam_r         250  cam_comp.F90
2546: e3sm.exe           00000000004D9CAC  atm_comp_mct_mp_a         341  atm_comp_mct.F90
2546: e3sm.exe           000000000042367F  component_mod_mp_         269  component_mod.F90
2546: e3sm.exe           000000000041768D  cime_comp_mod_mp_        1958  cime_comp_mod.F90
2546: e3sm.exe           000000000042042C  MAIN__                     92  cime_driver.F90
2546: e3sm.exe           0000000000401ADE  Unknown               Unknown  Unknown
2546: e3sm.exe           000000000440968F  Unknown               Unknown  Unknown
2546: e3sm.exe           00000000004019CA  Unknown               Unknown  Unknown
`

**Comment by tangq** [comment]
@singhbalwinder , if this floating invalid error is harmless and removing the debug flags in the default compile won't change the results, can you provide some instructions on how to remove those debug flags? In this way, I can start this time sensitive simulation now.

We can verify if the results change by comparing with the old results from the same settings. Thanks!

**Comment by singhbalwinder** [comment]
That is a good plan. I looked into `/global/cscratch1/sd/tang30/E3SM_simulations/cori-knl.20191124.tst2.F2010C5-CMIP6-LR.ne30_oECv3` but I couldn't find the build log files there in the `build` directory. I would like to look at them to understand how the flags are being selected.

If you can work within the `case_scripts` directory, please follow the following steps:

1. Edit `Macros.make` file to remove fpe0 from the following block:
```
ifeq ($(DEBUG), FALSE)
   FFLAGS +=  -O2 -debug minimal -qno-opt-dynamic-align -fpe0
   CFLAGS +=  -O2 -debug minimal
endif
```
2. Remove the `build` directory and run the following commands:

`./case.build --clean-all` (this is not needed if you remove the build directory but it doesn't hurt)
`./case.build`
`./case.submit`

Now this will let the model run without crashing with a floating invalid error but it is very dangerous to run the model this way when we know there is an issue (probably you already know this). I will investigate in the meanwhile. Sometimes these errors can be innocuous but they can be answer changing as well. Here we are working under the assumption that fixing this error will not cause the answers to change.

**Comment by tangq** [comment]
@singhbalwinder , you can find the building log here: `/global/cscratch1/sd/tang30/E3SM_simulations/cori-knl.20191124.F2010C5-CMIP6-LR.ARE.nudgeUV.ne30_oECv3/build`. I reused the executable as both tests use the same code but different user_nl settings.

I checked the building log. The `-fpe0` flag was applied.

**Comment by tangq** [comment]
I am also testing the default build on cori-haswell to see if this issue is machine specific. However, the haswell queue is too crowded - hopefully the test can start overnight.

**Comment by singhbalwinder** [comment]
Thanks @tangq . After your new build with the changed Macros.make file, please look at the build logs again to ensure that `-fpe0` is not used while building the model.

I just checked and found the following in the machine files of maint-1.0:
```
<compiler COMPILER="intel" MACH="cori-knl">
  <FFLAGS> -convert big_endian -assume byterecl -ftz -traceback -assume realloc_lhs -fp-model consistent -fimf-use-svml </FFLAGS>
  <ADD_FFLAGS DEBUG="FALSE"> -O2 -debug minimal -qno-opt-dynamic-align -fpe0</ADD_FFLAGS>
```
This is the only place I found `-fpe0` is being used when `DEBUG=FLASE`. Unless there is a good reason to keep it, we should remove it as it does generally slows down the simulations.

I don't think running on Haswell will help much as it may either confirm the issue or may just run fine. Cori-Haswell doesn't have this flag for DEBUG=FALSE, so it might just run fine.

**Comment by tangq** [comment]
I did confirm that `-fpe0` was not used in the atm build log. Thanks! @singhbalwinder 

The test is waiting in the queue. Let's see how it goes tomorrow. Good night.

**Comment by singhbalwinder** [comment]
@tangq : I ran this case with DEBUG=TRUE on Compy and it ran fine till 66th time step and after that it died due to the time limit. At which time step do you see this floating invalid error? Perhaps only Cori-KNL can reproduce this error, I will try it there.

**Comment by tangq** [comment]
@singhbalwinder , my error occurred during initialization. So, your compy test seems doesn't have this error.

@wlin7 did a test on cori-knl with the default non-DEBUG build and did NOT have the issue. I am comparing mine with Wuyin's to see if what are the differences between our two tests.

**Comment by tangq** [comment]
I only found one difference between my test and @wlin7 's in drv_in
```
118c118
<   do_budgets = .false.
---
>   do_budgets = .true.
```
Will this difference cause floating invalid?

**Comment by ndkeen** [comment]
I agree about the `-fpe0` flag. It was not added by me, but since it was used for some official runs, we didn't want to change it unless needed. Master doesn't use this flag for non-debug runs. 

I tried the following 2 tests with maint-1.0 from Oct 23rd (is the date of the repo important?)

```
SMS_PT.ne30_oECv3.F2010C5-CMIP6-LR
SMS_PL.ne30_oECv3.F2010C5-CMIP6-LR
```

The smaller node count run completed. The Large run failed in the same way as you noted. I also tried building again without `-fpe0` and I see same result. Looking at the code, that is the exact line where we were finding uninit issues with DEBUG=TRUE in master.  The initialization of qrl_idx seemed to work (which is in the PR noted above). I'm trying to build with that now. Qi noted that it fails in a different place when he tried.

This still might be the first time I've seen the code fail this way with DEBUG=FALSE.

**Comment by wlin7** [comment]
> Will this difference cause floating invalid? (do_budgets = false or true)

No, it has no impact on my test runs. With do_budgets = TRUE of FALSE, the runs are BFB.

**Comment by rljacob** [comment]
It won't change answers but do_budgets = true will kick off a bunch of floating point calculations that otherwise won't happen.

**Comment by singhbalwinder** [comment]
If the failure is node count dependent, it is most likely a memory issue (e.g. uninitialized variable). I will see if I can reproduce it on Compy. My account on Cori has some issues which I am trying to sort out.

**Comment by tangq** [comment]
@wlin7 and I tested 1) do_budgets=true 2) the default 41-node layout (the problematic one uses  81 nodes `--pecount L`) 3) do_budgets=true and 41-node layout.

Test 1) failed, while the other two were successful. So, it seems that the failure is node count dependent. @wlin7 noticed that ne30 has 5400 elements and the failing run has 5427 MPI tasks. That is more MPI tasks than # of elements.

I will use the default layout for my F-cases on cori-knl, so the simulations won't be delayed.

The tests are located under /global/cscratch1/sd/tang30/E3SM_simulations, in case they are useful for the fix.
Test 1: cori-knl.20191124.tstBudget.F2010C5-CMIP6-LR.ne30_oECv3
Test 2: cori-knl.20191124.tstlayout.F2010C5-CMIP6-LR.ne30_oECv3
Test 3: cori-knl.20191124.tstlayoutBudget.F2010C5-CMIP6-LR.ne30_oECv3

**Comment by singhbalwinder** [comment]
Thanks @tangq. These were interesting finds. I am tagging @mt5555 here as we have seen something similar before where the model would break if we use more tasks than the model elements. I may be wrong but I think @mt5555 fixed that so that the model gives a nice error message. May be the fix is applied to the master but not to the maint-1.0 branch. In any case, I feel more comfortable going ahead with the simulations after these findings.

**Comment by singhbalwinder** [comment]
I think that PR was #3064.

**Comment by tangq** [comment]
@singhbalwinder , it's worth mentioning that my previous floating invalid error we talked about is also on cori-knl with the same 81-node layout. The error occurred with certain `pertlim` numbers which are used to add white noise to the initial T field.

**Comment by ndkeen** [comment]
I think the code should be able to handle more MPI's than elements (which usually adds a slight benefit, but in this case 5427 was chosen only to fully use the cores on all 81 nodes).  And I just tried changing the PE layout to use 5400 tasks for all components that were using 5427 and I see the same error. This layout does not use threads.

**Comment by singhbalwinder** [comment]
I did some short tests with ne4 using more MPI tasks than elements. For example, ne4 has 96 elements:
` total number of elements nelem =           96`
When I run with Debug flags turned on, the simulation always crashes if I use `MPI tasks> nelem` with the following error:
```
 forrtl: severe (408): fort: (7): Attempt to use pointer ELEM when it is not associated with a target
96:
96: Image              PC                Routine            Line        Source
96: e3sm.exe           0000000008DFBB96  Unknown               Unknown  Unknown
96: e3sm.exe           000000000254A1B2  dyn_comp_mp_dyn_i         189  dyn_comp.F90
96: e3sm.exe           0000000000A84D53  inital_mp_cam_ini          31  inital.F90
96: e3sm.exe           000000000083E7DF  cam_comp_mp_cam_i         159  cam_comp.F90
96: e3sm.exe           000000000080A4E9  atm_comp_mct_mp_a         260  atm_comp_mct.F90
96: e3sm.exe           0000000000466454  component_mod_mp_         269  component_mod.F90
96: e3sm.exe           0000000000428B8A  cime_comp_mod_mp_        1174  cime_comp_mod.F90
96: e3sm.exe           000000000045C027  MAIN__                     92  cime_driver.F90
96: e3sm.exe           0000000000417322  Unknown               Unknown  Unknown
96: libc-2.17.so       00002B2607CC03D5  __libc_start_main     Unknown  Unknown
96: e3sm.exe           0000000000417229  Unknown               Unknown  Unknown
```

The simulation runs fine if `MPI tasks<= nelem`.

**Comment by ndkeen** [comment]
I also found this a while ago here:  https://github.com/E3SM-Project/E3SM/issues/2899

**Comment by qinyia** [comment]
I also met the same floating invalid error of F2010C5 on cori-knl with processor_config "L". Is there any update about fixing this problem?

**Comment by ndkeen** [comment]
Using most recent maint-1.0, I ran `SMS_D_Ln9_PL.ne30_oECv3.F2010C5-CMIP6` as well as `SMS_Ln9_PL.ne30_oECv3.F2010C5-CMIP6` and I get what looks like the same error as originally reported.

Which is not terribly useful, but just reporting we still see the same issue. I don't know if this is problematic or not (ie if this *should* be allowed to run).

```
2680: forrtl: error (65): floating invalid
2680: Image              PC                Routine            Line        Source             
2680: e3sm.exe           00000000041A386E  Unknown               Unknown  Unknown
2680: e3sm.exe           0000000003A3C240  Unknown               Unknown  Unknown
2680: e3sm.exe           000000000107D583  clubb_intr_mp_clu        1584  clubb_intr.F90
2680: e3sm.exe           00000000005FA3DB  physpkg_mp_tphysb        2480  physpkg.F90
2680: e3sm.exe           00000000005F6B12  physpkg_mp_phys_r        1036  physpkg.F90
2680: e3sm.exe           00000000004E79E7  cam_comp_mp_cam_r         250  cam_comp.F90
2680: e3sm.exe           00000000004D9CAC  atm_comp_mct_mp_a         341  atm_comp_mct.F90
2680: e3sm.exe           000000000042367F  component_mod_mp_         269  component_mod.F90
2680: e3sm.exe           000000000041768D  cime_comp_mod_mp_        1958  cime_comp_mod.F90
2680: e3sm.exe           000000000042042C  MAIN__                     92  cime_driver.F90
2680: e3sm.exe           0000000000401ADE  Unknown               Unknown  Unknown
```
